cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -83,19 +90,52 @@ public PushAppResponse push(PushAppOperationRequest pushRequest) throws NoChange
             pushRequest.getNamespaceId(),
             pushRequest.getRepositoryConfig())
     ) {
-      try {
-        repositoryManager.cloneRemote();
-      } catch (GitAPIException | IOException e) {
-        throw new GitOperationException(String.format("Failed to clone remote repository: %s",
-                                                       e.getMessage()), e);
-      }
+      cloneAndCreateBaseDir(repositoryManager);
+
+      writeAppDetail(repositoryManager, pushRequest.getApp());
+
+      // it should never be empty as in case of any error we will get an exception
+      return commitAndPush(
+          repositoryManager,
+          ImmutableSet.of(new AppVersion(
+              pushRequest.getApp().getName(), pushRequest.getApp().getAppVersion()
+          )),
+          pushRequest.getCommitDetails()
+      ).get(0);
+    }
+  }
+
+  @Override
+  public List<PushAppResponse> multiPush(MultiPushAppOperationRequest pushRequest,
+      ApplicationManager appManager)
+      throws NoChangesToPushException, AuthenticationConfigException {
+    try (
+        RepositoryManager repositoryManager = repoManagerFactory.create(
+            pushRequest.getNamespaceId(),
+            pushRequest.getRepositoryConfig())
+    ) {
+      cloneAndCreateBaseDir(repositoryManager);
+
+      LOG.info("Pushing application configs for : {}", pushRequest.getApps());
 
-      LOG.info("Pushing application configs for : {}", pushRequest.getApp().getName());
+      Set<AppVersion> appVersions = new HashSet<>();
 
-      //TODO: CDAP-20371, Add retry logic here in case the head at remote moved while we are doing push
-      return writeAppDetailAndPush(
+      for (String appToPush : pushRequest.getApps()) {
+        ApplicationReference appRef = new ApplicationReference(pushRequest.getNamespaceId(),
+            appToPush);
+        try {
+          ApplicationDetail detail = appManager.get(appRef);
+          writeAppDetail(repositoryManager, detail);
+          appVersions.add(new AppVersion(detail.getName(), detail.getAppVersion()));
+        } catch (IOException | NotFoundException e) {
+          throw new SourceControlException(
+              String.format("Failed to fetch details for app %s", appRef));
```

Comment:
Always retain the original exception as the cause when rethrowing.

Reply:
updated.